src/main/java/org/opentripplanner/updater/car_rental/CarRentalUpdater.java

```diff
@@ -178,23 +183,40 @@ private void applyStations(Graph graph) {
                 CarRentalStation station = entry.getKey();
                 if (stationSet.contains(station))
                     continue;
-                CarRentalStationVertex vertex = entry.getValue();
-                if (graph.containsVertex(vertex)) {
-                    graph.removeVertexAndEdges(vertex);
-                }
+
+                removeStationVerticesAndEgesFromGraph(entry.getValue());
+
                 toRemove.add(station);
                 service.removeCarRentalStation(station);
-                // TODO: need to unsplit any streets that were split
             }
             for (CarRentalStation station : toRemove) {
                 // post-iteration removal to avoid concurrent modification
                 verticesByStation.remove(station);
             }
         }
 
+        private void removeStationVerticesAndEgesFromGraph(CarRentalStationVertex carRentalStationVertex) {
+            // before removing the carRentalStationVertex, first find and remove all associated
+            // SemiPermanentSplitterVertices
+            for (Edge edge : carRentalStationVertex.getOutgoing()) {
+                if (edge instanceof StreetCarRentalLink) {
+                    StreetCarRentalLink toStreetLink = (StreetCarRentalLink) edge;
+                    StreetVertex streetVertex = (StreetVertex) toStreetLink.getToVertex();
+                    if (streetVertex != null && streetVertex instanceof SemiPermanentSplitterVertex) {
+                        splitter.removeSemiPermanentVerticesAndEdges((SemiPermanentSplitterVertex) streetVertex);
+                    }
+                }
+            }
+
+            // remove the carRentalStationVertex from the graph if it's in there (why wouldn't it be?)
+            if (graph.containsVertex(carRentalStationVertex)) {
+                graph.removeVertexAndEdges(carRentalStationVertex);
+            }
+        }
```

Comment:
Is there an opportunity to combine this code with the bike rental updater method? Just wondering.

Reply:
For sure, but maybe this is better done in the great vehicle rental refactor that will consolidate all of the vehicle rental code. There are still some mode-specific vertex and edge types for now that I think makes sense to leave as is.